File: bindings/javascript/src/modbus.js
// SPDX-License-Identifier: PMPL-1.0-or-later
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <j.d.a.jewell@open.ac.uk>
//
// Modbus protocol types for proven-servers.

/** FunctionCode matching the Idris2 ABI tags. */
export const FunctionCode = Object.freeze({
  READ_COILS: 0,
  READ_DISCRETE_INPUTS: 1,
  READ_HOLDING_REGISTERS: 2,
  READ_INPUT_REGISTERS: 3,
  WRITE_SINGLE_COIL: 4,
  WRITE_SINGLE_REGISTER: 5,
  WRITE_MULTIPLE_COILS: 6,
  WRITE_MULTIPLE_REGISTERS: 7,
  READ_WRITE_MULTIPLE_REGISTERS: 8,
  MASK_WRITE_REGISTER: 9,
});

/** ExceptionCode matching the Idris2 ABI tags. */
export const ExceptionCode = Object.freeze({
  ILLEGAL_FUNCTION: 0,
  ILLEGAL_DATA_ADDRESS: 1,
  ILLEGAL_DATA_VALUE: 2,
  SLAVE_DEVICE_FAILURE: 3,
  ACKNOWLEDGE: 4,
  SLAVE_DEVICE_BUSY: 5,
  MEMORY_PARITY_ERROR: 6,
  GATEWAY_PATH_UNAVAILABLE: 7,
  GATEWAY_TARGET_DEVICE_FAILED: 8,
});

/** DeviceRole matching the Idris2 ABI tags. */
export const DeviceRole = Object.freeze({
  MASTER: 0,
  SLAVE: 1,
});

/** GatewayState matching the Idris2 ABI tags. */
export const GatewayState = Object.freeze({
  IDLE: 0,
  LISTENING: 1,
  PROCESSING: 2,
  ERROR: 3,
  STOPPING: 4,
});

File: bindings/dart/lib/src/modbus.dart
// SPDX-License-Identifier: PMPL-1.0-or-later
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <j.d.a.jewell@open.ac.uk>
//
// Modbus protocol types for proven-servers.

/// FunctionCode matching the Idris2 ABI tags.
enum FunctionCode {
  readCoils(0),
  readDiscreteInputs(1),
  readHoldingRegisters(2),
  readInputRegisters(3),
  writeSingleCoil(4),
  writeSingleRegister(5),
  writeMultipleCoils(6),
  writeMultipleRegisters(7),
  readWriteMultipleRegisters(8),
  maskWriteRegister(9);

  const FunctionCode(this.tag);
  final int tag;

  static FunctionCode? fromTag(int tag) {
    for (final v in values) {
      if (v.tag == tag) return v;
    }
    return null;
  }
}

/// ExceptionCode matching the Idris2 ABI tags.
enum ExceptionCode {
  illegalFunction(0),
  illegalDataAddress(1),
  illegalDataValue(2),
  slaveDeviceFailure(3),
  acknowledge(4),
  slaveDeviceBusy(5),
  memoryParityError(6),
  gatewayPathUnavailable(7),
  gatewayTargetDeviceFailed(8);

  const ExceptionCode(this.tag);
  final int tag;

  static ExceptionCode? fromTag(int tag) {
    for (final v in values) {
      if (v.tag == tag) return v;
    }
    return null;
  }
}

/// DeviceRole matching the Idris2 ABI tags.
enum DeviceRole {
  master(0),
  slave(1);

  const DeviceRole(this.tag);
  final int tag;

  static DeviceRole? fromTag(int tag) {
    for (final v in values) {
      if (v.tag == tag) return v;
    }
    return null;
  }
}

/// GatewayState matching the Idris2 ABI tags.
enum GatewayState {
  idle(0),
  listening(1),
  processing(2),
  error(3),
  stopping(4);

  const GatewayState(this.tag);
  final int tag;

  static GatewayState? fromTag(int tag) {
    for (final v in values) {
      if (v.tag == tag) return v;
    }
    return null;
  }
}
